lib/Dialect/FIRRTL/FIRRTLAnnotationHelper.cpp

```diff
@@ -333,21 +333,18 @@ std::optional<AnnoPathValue> firrtl::resolvePath(StringRef rawPath,
   return resolveEntities(*tokens, circuit, symTbl, cache);
 }
 
-InstanceOp firrtl::addPortsToModule(
-    FModuleLike mod, InstanceOp instOnPath, FIRRTLType portType, Direction dir,
-    StringRef newName, InstancePathCache &instancePathcache,
-    llvm::function_ref<ModuleNamespace &(FModuleLike)> getNamespace,
-    CircuitTargetCache *targetCaches) {
+InstanceOp firrtl::addPortsToModule(FModuleLike mod, InstanceOp instOnPath,
+                                    FIRRTLType portType, Direction dir,
+                                    StringRef newName,
+                                    InstancePathCache &instancePathcache,
+                                    CircuitTargetCache *targetCaches) {
   // To store the cloned version of `instOnPath`.
   InstanceOp clonedInstOnPath;
   // Get a new port name from the Namespace.
-  auto portName = [&](FModuleLike nameForMod) {
-    return StringAttr::get(nameForMod.getContext(),
-                           getNamespace(nameForMod).newName(newName));
-  };
+  auto portName = StringAttr::get(mod.getContext(), newName);
   // The port number for the new port.
   unsigned portNo = getNumPorts(mod);
-  PortInfo portInfo = {portName(mod), portType, dir, {}, mod.getLoc()};
+  PortInfo portInfo = {portName, portType, dir, {}, mod.getLoc()};
   mod.insertPorts({{portNo, portInfo}});
   if (targetCaches)
     targetCaches->insertPort(mod, portNo);
```

Comment:
Hmm.  `targetCaches` contains `AnnoTargetCache` per-module, and is used for resolving annotations.
This code in theory allows the new port to be used as target of an annotation, possibly unprocessed but probably more reasonably as part of the "this annotation adds other annotations, add to worklist" pattern.

In practice, port names are/must (I believe) be unique after parsing, so this is nearly right, but the port-unique-check should be checking the other names not symbols.

As you pointed out, the current code does help when repeatedly adding port of same "newName", but seems otherwise wrong re:not avoiding conflicts with existing ports.

It seems right we keep the target cache up-to-date (or explicitly don't).  Not sure how to best right this offhand.

Reply:
Yeah, I think its possibly a mistake to be adding these ports to the target cache, which allows people to have annotations target ports which are added for GrandCentral or WiringTransform. The only valid use-case I can think of is chaining internal-only annotations but I have a feeling we don't do that. It wouldn't be too hard to make sure the new port has a unique name if we must. @seldridge any opinion here?

Reply:
FWIW I was having trouble tracking down if we are chaining annotations as mentioned above. I locally removed the AnnoTargetCache updates from this function as well, which resulted no changes to the RTL.

Reply:
All calls to `addToWorklistFn` are the sort of "chaining" I meant, but if you mean "create something new and expect to be able to add an annotation to it this way"... I think this maybe isn't a thing ("anymore")?  I believe SignalDriving (GCSM) added ports everywhere and then wanted to annotate them, or so.

We do need to kinda keep the cache up-to-date re:replacing ops (instances in particular), FWIW looks like the only user of this method is for legacy wiring problem implementation, which already has a fun issue re:invalidating instance refs.  See https://github.com/llvm/circt/pull/4496#discussion_r1065935380 and linked from there for more info.

re:no changes to RTL, we don't use legacy wiring (and maybe soon maybe not firtool-based wiring at all-- although don't think can drop entirely anytime soon) so not surprising changing this code had no effect.

That single user, however, adds new reftype ports and those definitely cannot/shouldn't be targeted by annotations chained or otherwise.

All of this is to say, I think we can rework this to do what your PR does (and move us in the direction we're trying to go that motivated it) with a bit of reworking.

I think the targetCache should be initialized once and only "updated" if you're replacing/changing things it knows about, no new IR should go there as it's not valid for target by annotation.

Reply:
Whoops, this interacts with legacy wiring but appears to be used by the `internalPath` thing instead.

Reply:
So we decided that we should not be adding the newly added port to the AnnoTargetCache, since it should not be targetable by annotations.